pandas/_libs/lib.pyx

```diff
@@ -571,6 +571,11 @@ def array_equivalent_object(left: object[:], right: object[:]) -> bool:
             if PyArray_Check(x) and PyArray_Check(y):
                 if not array_equivalent_object(x, y):
                     return False
+            elif x is C_NA or y is C_NA:
```

Comment:
I think you can just simplify by doing an XOR and returning False when that is the case. PyObject_RichCompareBool on the next condition will return True when both objects have the same id anyway

Reply:
Ah yes, good idea